### PR TITLE
KIT-676 Include analytics action in thunk payload

### PR DESCRIPTION
**Pull request by samisayegh**
_This story was an exploration, aiming to give developers building with headless a more consistent experience. I'm not sure if it's right to merge, but I wanted to share what I learned and get feedback._

I opened [this jira](https://coveord.atlassian.net/browse/KIT-676) when working on the quickview feature, after seeing other async thunks were bundling the analytics event inside the async action.

**Pros**
I thought this was quite handy. A benefit is if someone dispatched the action directly instead of using the controller, the analytics event will still be logged. This removes the chance of someone forgetting to log. Also, if the api ever changed (e.g. moving to collect), we can handle the migration internally.

**Cons**
One disadvantage is it is harder to test. I needed to use a cypress test instead of a unit test. Not too bad.
However, this revealed a deeper problem. Even though the test passes, the UA response is actually a 400, because the event is missing a search uid. Related, logging analytics means the `fetchResultContent` action needs a valid `Result` object, instead of just a `uniqueId`.

**Take aways**
There is a disconnect between what is needed by the search api to retrieve content, and what is needed by analytics.
If when switching apis, `collect` requires additional/different data, it would mean a breaking change, even though search api is still the same.
If a use-case just wants to retrieve html by `uniqueId`, now they need to pass a `Result` object which is less convenient.

**tl;dr**
1. Not bundling analytics inside actions provides more flexibility, by avoiding UA errors, simplifying parameters.
2. Bundling analytics is more opinionated, forces developers to do things a certain way (pass `Result` object), but allows us to manage UA behind the scenes.

https://coveord.atlassian.net/browse/KIT-676

### CODE REVIEW

File: packages/atomic/cypress/integration/quickview.cypress.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      const analyticsBody = request.body;
+      expect(analyticsBody).to.have.property(
+        'actionCause',
+        'documentQuickview'
+      );
+    });
+  });
+});
+
+function buildResult(config: Partial<Result>): Result {
```
Comment (ThibodeauJF) [inline packages/atomic/cypress/integration/quickview.cypress.ts:43]: Do we have exported `TestUtils` from Headless that we can use?

**Review by ThibodeauJF — APPROVED** [review]
I think the pros out-weight the cons, could we easily resolve the result object using the uniqueId from the state, or am I forgetting something?

**Comment by btaillon** [comment]
I'm not sure how I feel about having analytics sometimes be a side-effect of an action and sometimes be manual.

Additionally, I don't think we should expect too many developers to be using actions instead of controllers. If they decide to use actions instead of controllers, I think the responsibility could fall on them to update analytics when they update the version of headless.

**Comment by samisayegh** [comment]
> Could we easily resolve the result object using the uniqueId from the state

We could, in this case we are assuming that the dev executed a search first, and if not (as in the cypress test atm), it would fail when logging an analytics event.

It's the question of whether we want the action to behave like an api client, or as part of an opinionated system, where to use quickview actions, you must first have performed a search that returned a result with a `uniqueId` matching the one being quickviewed.

> I'm not sure how I feel about having analytics sometimes be a side-effect of an action and sometimes be manual.

Same, ideally it would be one way or the other, with the exception being if the action explicitly asks for an analytics action to log (e.g. `executeSearch`).

I think the idea of letting developers log analytics themselves if they are working with actions directly makes a lot of sense.
It's more flexible, but loses out on certain pros,
